File: client-examples/python/null_auth_sdk.py
#!/usr/bin/env python3
"""
Null-Auth Core Client SDK Library
Zero-dependency Python 3 client for Null-Auth Private Licensing & Auth Platform.
Supports BOTH Auth Methods:
  1. License Key Authentication (License Key + Bound Machine SID)
  2. HWID Whitelist Authentication (Direct Machine SID Authorization)
"""

import subprocess
import json
import urllib.request
import urllib.error

class NullAuthSDK:
    def __init__(self, app_id: str, app_secret: str, base_url: str = "https://null-auth-backend.vercel.app"):
        self.base_url = base_url.rstrip('/')
        self.app_id = app_id
        self.app_secret = app_secret

    @staticmethod
    def get_windows_user_sid() -> str:
        """
        Retrieves the Windows User SID using standard 'whoami /user' command.
        """
        try:
            output = subprocess.check_output(["whoami", "/user"], text=True)
            for line in output.splitlines():
                if "S-1-5-" in line:
                    parts = line.split()
                    if parts:
                        return parts[-1].strip()
        except Exception:
            pass
        return "UNKNOWN_HWID"

    def authenticate_license(self, license_key: str) -> dict:
        """
        METHOD 1: License Key Authentication
        Validates license key and binds to Windows User SID.
        """
        url = f"{self.base_url}/api/v1/client/license/authenticate"
        payload = {
            "appId": self.app_id,
            "appSecret": self.app_secret,
            "licenseKey": license_key.strip(),
            "hwid": self.get_windows_user_sid()
        }
        return self._send_request(url, payload)

    def authenticate_hwid(self) -> dict:
        """
        METHOD 2: HWID Whitelist Authentication
        Direct machine-level authorization without requiring a license key.
        """
        url = f"{self.base_url}/api/v1/client/hwid/authenticate"
        payload = {
            "appId": self.app_id,
            "appSecret": self.app_secret,
            "hwid": self.get_windows_user_sid()
        }
        return self._send_request(url, payload)

    def _send_request(self, url: str, payload: dict) -> dict:
        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json", "User-Agent": "NullAuthPythonSDK/1.0"},
            method="POST"
        )
        try:
            with urllib.request.urlopen(req) as response:
                return json.loads(response.read().decode('utf-8'))
        except urllib.error.HTTPError as e:
            try:
                return json.loads(e.read().decode('utf-8'))
            except Exception:
                return {"success": False, "message": f"HTTP Error {e.code}"}
        except Exception:
            return {"success": False, "message": "Failed to connect to Null-Auth cloud server."}

File: client-examples/csharp/NullAuthSDK.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Security.Principal;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace NullAuthClient
{
    public class NullAuthResult
    {
        public bool Success { get; set; }
        public string Message { get; set; }
        public string ErrorCode { get; set; }
        public NullAuthData Data { get; set; }
    }

    public class NullAuthData
    {
        public string Status { get; set; }
        public string ExpiresAt { get; set; }
        public int RemainingDays { get; set; }
        public string FirstActivatedAt { get; set; }
    }

    /// <summary>
    /// Core Null-Auth SDK Library implementing BOTH Auth Methods:
    ///   - Method 1: License Key Authentication
    ///   - Method 2: Direct HWID Whitelist Authentication
    /// </summary>
    public class NullAuthSDK
    {
        private readonly string _baseUrl;
        private readonly string _appId;
        private readonly string _appSecret;
        private static readonly HttpClient _httpClient = new HttpClient();

        public NullAuthSDK(string appId, string appSecret, string baseUrl = "https://null-auth-backend.vercel.app")
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _appId = appId;
            _appSecret = appSecret;
        }

        /// <summary>
        /// Retrieves the Windows User SID using legitimate 'whoami /user' command.
        /// </summary>
        public static string GetWindowsUserSid()
        {
            try
            {
                ProcessStartInfo psi = new ProcessStartInfo("whoami", "/user")
                {
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                };

                using (Process process = Process.Start(psi))
                {
                    string output = process.StandardOutput.ReadToEnd();
                    process.WaitForExit();

                    string[] lines = output.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
                    foreach (string line in lines)
                    {
                        if (line.Contains("S-1-5-"))
                        {
                            string[] parts = line.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                            if (parts.Length >= 2)
                            {
                                return parts[parts.Length - 1].Trim();
                            }
                        }
                    }
                }
            }
            catch { }

            try
            {
                return WindowsIdentity.GetCurrent().User?.Value ?? "UNKNOWN_SID";
            }
            catch
            {
                return "UNKNOWN_HWID";
            }
        }

        /// <summary>
        /// METHOD 1: Authenticate using License Key (License Key + Bound Machine SID)
        /// </summary>
        public async Task<NullAuthResult> AuthenticateLicenseAsync(string licenseKey)
        {
            string hwid = GetWindowsUserSid();
            string endpoint = $"{_baseUrl}/api/v1/client/license/authenticate";

            var payload = new
            {
                appId = _appId,
                appSecret = _appSecret,
                licenseKey = licenseKey?.Trim(),
                hwid = hwid
            };

            return await SendAuthRequestAsync(endpoint, payload);
        }

        /// <summary>
        /// METHOD 2: Authenticate using HWID Whitelist directly
        /// </summary>
        public async Task<NullAuthResult> AuthenticateHwidAsync()
        {
            string hwid = GetWindowsUserSid();
            string endpoint = $"{_baseUrl}/api/v1/client/hwid/authenticate";

            var payload = new
            {
                appId = _appId,
                appSecret = _appSecret,
                hwid = hwid
            };

            return await SendAuthRequestAsync(endpoint, payload);
        }

        private async Task<NullAuthResult> SendAuthRequestAsync(string endpoint, object payload)
        {
            try
            {
                string jsonBody = JsonSerializer.Serialize(payload);
                var content = new StringContent(jsonBody, Encoding.UTF8, "application/json");

                HttpResponseMessage response = await _httpClient.PostAsync(endpoint, content);
                string responseString = await response.Content.ReadAsStringAsync();

                using JsonDocument doc = JsonDocument.Parse(responseString);
                JsonElement root = doc.RootElement;

                bool success = root.GetProperty("success").GetBoolean();
                string message = root.GetProperty("message").GetString();
                string errorCode = root.TryGetProperty("error", out var errElem) ? errElem.GetString() : null;

                NullAuthData data = null;
                if (success && root.TryGetProperty("data", out var dataElem))
                {
                    data = new NullAuthData
                    {
                        Status = dataElem.TryGetProperty("status", out var s) ? s.GetString() : "active",
                        ExpiresAt = dataElem.TryGetProperty("expires_at", out var e) ? e.GetString() : "",
                        RemainingDays = dataElem.TryGetProperty("remaining_days", out var r) ? r.GetInt32() : 0,
                        FirstActivatedAt = dataElem.TryGetProperty("first_activated_at", out var f) ? f.GetString() : null,
                    };
                }

                return new NullAuthResult
                {
                    Success = success,
                    Message = message,
                    ErrorCode = errorCode,
                    Data = data
                };
            }
            catch (Exception ex)
            {
                return new NullAuthResult
                {
                    Success = false,
                    Message = "Connection failed to Null-Auth server.",
                    ErrorCode = "NETWORK_ERROR",
                    Data = null
                };
            }
        }
    }
}
